### ISSUE TITLE
Stability issues when resuming from diff snapshots

### ISSUE THREAD

**Issue by vadimberezniker** [firecracker-microvm/firecracker#2802]
# Describe the bug

We have been experimenting with snapshot support in Firecracker, but have been running into some stability issues when booting from diff snapshots. We have not experienced any issues when using only full snapshots.

The "stability" issues appear to be symptoms of memory corruption, for example unexpected kernel panics, weird application segfaults and otherwise strange application behavior. 

The machine I'm testing on:
`Linux aorus 5.11.0-40-generic #44~20.04.2-Ubuntu SMP Tue Oct 26 18:07:44 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
I've also tested this on different Intel-based machine. 

## To Reproduce

1. Start a VM configured with a vsock device, networking and a block device (not sure if any of these datails matter, but including them for completeness)
2. Run the following commands:
```
git clone https://github.com/buildbuddy-io/buildbuddy.git
cd buildbuddy
bazelisk build //flag
```
note: the bazelisk command will ultimately error out because there's no //flag target but that's expected. 
2. Pause VM
3. Take a full snapshot
4. Remove VM
5. Create new VM from full snapshot
6. Run `bazelisk build //flag`
7. Pause VM
8. Take diff snapshot
9. Merge diff snapshot into full snapshot from step (3). I've tried using both custom code and a `dd bs=4096 if=diff of=full conv=sparse,notrunc` command with same results. 
10. Remove VM
11. Create new VM from full memory snapshot made in step (9) and state snapshot made in step (8)
12. Run `bazelisk build //flag`

## Expected behavior

Output of step 12 should be the same as output of step 6 and should complete relatively quickly. 
What we actually see here is strange behaviors such as kernel panics, strange segfaults and otherwise unexpected guest behavior. 
Everything is fine if we replace step 8-9 with another full memory snapshot instead of a diff snapshot.

It's worth noting I have also done a test after step 7 where i take a diff snapshot followed by full snapshot to compare the full snapshot and the generated merged snapshot and verified that they are identical. 

## Environment

`Firecracker v0.25.1`
`Host kernel: 5.11.0-40-generic, guest kernel: 5.4.0+`
`Arch: x86_64`
`CPU: AMD Ryzen 9 3950X 16-Core Processor`
(Note I've also tested on an Intel-based machine.)

FS information:
The root FS we are using is available here: https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/vmsupport/bin/initrd.cpio
It contains our init binary: https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/server/cmd/goinit/main.go
The init process mounts a FS made from an OCI image and makes that the new root. 
The OCI image I am using for testing is gcr.io/flame-public/buildbuddy-ci-runner@sha256:fbff6d1e88e9e1085c7e46bd0c5de4f478e97b630246631e5f9d7c720c968e2e
It includes the git, bazelisk, etc commands referenced in the repro steps. 

## Additional context

We are trying to make use of diff snapshots to improve snapshot performance as diff snapshots can be significantly faster than always taking full snapshots. 

Please let me know if I've missed anything. I've tried my best to list everything that might be relevant to the issue at hand.

## Checks

- [x] Have you searched the Firecracker Issues database for similar problems?
- [x] Have you read the existing relevant Firecracker documentation?
- [kinda] Are you certain the bug being reported is a Firecracker issue?

**Comment by raduweiss** [comment]
@vadimberezniker thank you for reporting this. Just an additional question so we can accurately reproduce this: How do you handle block devices on steps (5) and (8)?

**Comment by vadimberezniker** [comment]
We use the same block device throughout. The same FS block file is used when creating the VMs in step (1), (5) and (11). 
The VM state always moves forward, we never try to create a VM using the current block device state but older snapshot.

**Comment by vadimberezniker** [comment]
> > 1. Start a VM configured with a vsock device, networking and a block device (not sure if any of these datails matter, but including them for completeness)
> 
> Can you add more details ? Memory size matters. IIRC if you have more than 3328MB Firecracker sets up 2 memory regions instead of one.
> 

I am testing with 5000MB VMs. I will try testing with a smaller memory configuration to see if the issues goes away.

> > It's worth noting I have also done a test after step 7 where i take a diff snapshot followed by full snapshot to compare the full snapshot and the generated merged snapshot and verified that they are identical.
> 
> This is interesting. There is an integration test that covers exactly this (sadly, with just one memory region): https://github.com/firecracker-microvm/firecracker/blob/main/tests/integration_tests/functional/test_snapshot_basic.py#L394 The entire scenario is also covered here. integration test: https://github.com/firecracker-microvm/firecracker/blob/main/tests/integration_tests/functional/test_snapshot_basic.py#L312
> 
> It is important to note that creating 2 snapshots in a row after a pause might produce different state files (only the first one is valid). IIRC some registers might reset after being saved here https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/vstate/vcpu/x86_64.rs#L250-L257
> 

I am always using the latest state snapshot when creating VM.

> @vadimberezniker You can always use Diff snapshots , even as base a snapshot, it's faster! There is a bit of confusion regarding the semantics of Full/Diff. In the context of the API "Full" just means that FC dump all write memory page by page to disk. There is not requirement for the first mem snapshot (base) to have the zeroes written - sparse file holes are backed by zero page on PF.
> 
> Would be great if you can reduce the complexity of your reproduction scenario - does it happen without your specific workload ? I would imagine we can reproduce it using the integration tests which should guard against regressions.

Let me see if I understand what you are saying. You are suggesting that instead of doing full snapshot -> diff -> diff -> diff ... we do diff -> diff -> diff -> diff because even if the type in the API is specified as diff the first snapshot should have all the non-zero pages written out so it should work just as well as a base snapshot. 

To summarize what I will try:
1) VM with smaller memory config
2) Creating base snapshot using 'diff' type

**Comment by vadimberezniker** [comment]
> 1. VM with smaller memory config

I tried creating VMs with 2500MB memory but ran into similar issues.

> 2. Creating base snapshot using 'diff' type

This improved the initial snapshot performance, thanks!

**Comment by vadimberezniker** [comment]
I think I was able to come up with a smaller repro here: https://github.com/vadimberezniker/fcsnapshotissue

To run the repro:
Compile `vmtest.go` and copy the binary onto the rootfs you are using. 
For my testing I used the rootfs & kernel linked from the Firecracker getting started guide.

Open one terminal and run `repro_vmm.sh`. This will create the firecracker process in a loop.
Open a second terminal and run `repro_client.sh Diff`. This script will perform snapshot + restore in a loop every 5 seconds. 
When the VM boots in the first terminal, run the `vmtest` binary. Then sit back and enjoy. 

In my testing after a few iterations (typically <10) I will see an error like:
```
[  106.319681] BUG: Bad page state in process vmtest  pfn:15e478
[  106.320119] page:ffffea0005791e00 count:0 mapcount:-1 mapping:          (null) index:0x1
[  106.320723] flags: 0x2fffe000000014(referenced|dirty)
[  106.321102] raw: 002fffe000000014 0000000000000000 0000000000000001 00000000fffffffe
[  106.321690] raw: dead000000000100 dead000000000200 0000000000000000 0000000000000000
[  106.322273] page dumped because: nonzero mapcount
```

When I ran the script with Full snapshots, it was able to go up to 150 iterations without any problems.

**Comment by serban300** [comment]
Hi @vadimberezniker 

Thank you for providing a simpler reproduction scenario ! It was super helpful in the investigation.

We managed to identify the root cause for this issue. The problem is that sometimes the dirtied pages are zeroed-out and in sparse files zeroed-out blocks are ignored. For example:

Let's say that we have a VM with 4 pages of memory and after the first snapshot we have the following memory layout and memory file:

| Page index | Page data |
|---|---|
| 1 | 1 |
| 2 | 2 |
| 3 | 3 |
| 4 | 4 |

We restore from this snapshot, run a workload and now the memory reaches the following state:

| Page index | Page data |
|---|---|
| 1 | 1 |
| 2 | 2 |
| 3 | 0 |
| 4 | 4 |

The KVM dirty log will look like this:

| Page index | is page dirty |
|---|---|
| 1 | false |
| 2 | false |
| 3 | true |
| 4 | false |

This will lead to the following diff file:

| Page index | Page data |
|---|---|
| 1 | 0 |
| 2 | 0 |
| 3 | 0 |
| 4 | 0 |

So even if the 3rd page is dirty, its value in the diff file will be 0. And since in sparse file 0s are being ignored, after running `dd`, the final merged file will look like the initial one:

| Page index | Page data |
|---|---|
| 1 | 1 |
| 2 | 2 |
| 3 | 3 |
| 4 | 4 |

But the correct final merged file should look like this:

| Page index | Page data |
|---|---|
| 1 | 1 |
| 2 | 2 |
| 3 | 0 |
| 4 | 4 |

And calling `debug.FreeOSMemory()` from `golang` causes memory pages to be reclaimed by the OS and zeroed out, triggering this exact scenario.

We're evaluating options for solving this issue. AFAICU ignoring 0s is a basic property of sparse files and I'm not sure if there's any way to get around this with the current approach.

**Comment by serban300** [comment]
@vadimberezniker 

I did some more research and it seems that when working with sparse files there is a difference between zeroed-out blocks and  holes. And the way Firecracker creates the diff file the zeroed-out pages are correctly written as blocks and not as holes. The problem is that `dd` simply ignores zeroes instead of ignoring holes when using `conv=sparse`.

I wrote a small program that merges files ignoring holes: https://github.com/serban300/merge_snap
It's just a test program and it's definitely not production ready. It's meant just to check the assumption stated above.
You have to compile it and you can run it like this : `merge_snap base_mem_file diff_mem_file`

Could you try to use it instead of `dd` and rerun the more complex reproduction scenario provided in the issue description ? Please let us know if it still reproduces.

**Comment by vadimberezniker** [comment]
Thanks a million! I'm implementing the equivalent fix in our code base and should have the results shortly!

**Comment by vadimberezniker** [comment]
I can confirm that using the proper sparse file APIs fixes the issues we were seeing with incremental snapshots.

**Comment by serban300** [comment]
Thanks for confirming this ! We'll keep the issue opened until we decide on an official solution for Firecracker.
